testbed/main.cc

```diff
@@ -207,27 +170,115 @@ int main(int argc, char* argv[])
 
   triangles = cdt->GetTriangles();
   map = cdt->GetMap();
-
-  cout << "Number of points = " << num_points << endl;
+  const size_t points_in_holes = std::accumulate(holes.cbegin(), holes.cend(), 0u,
+      [](size_t cumul, const Points& hole) { return cumul + hole.size(); });
+
+  cout << "Number of primary constrained edges = " << polyline.size() << endl;
+  cout << "Number of holes = " << holes.size() << endl;
+  cout << "Number of constrained edges in holes = " << points_in_holes << endl;
+  cout << "Number of Steiner points = " << steiner.size() << endl;
+  cout << "Total number of points = " << (polyline.size() + points_in_holes + steiner.size()) << endl;
   cout << "Number of triangles = " << triangles.size() << endl;
-  cout << "Elapsed time (ms) = " << dt*1000.0 << endl;
+  cout << "Elapsed time (ms) = " << dt * 1000.0 << endl;
 
   MainLoop(zoom);
 
   // Cleanup
-
   delete cdt;
-
-  // Free points
-  for(int i = 0; i < polylines.size(); i++) {
-    vector<Point*> poly = polylines[i];
-    FreeClear(poly);
+  FreeClear(polyline);
+  for (Points& hole : holes) {
+    FreeClear(hole);
   }
+  FreeClear(steiner);
 
   ShutDown(0);
   return 0;
 }
 
+bool ParseFile(string filename, Points& out_polyline, Polylines& out_holes, Points& out_steiner)
+{
+  enum ParserState
+  {
+    Polyline,
+    Hole,
+    Steiner,
+  };
+  ParserState state = Polyline;
+  Points* hole = nullptr;
+  try {
+    string line;
+    ifstream myfile(filename);
+    if (myfile.is_open()) {
+      while (!myfile.eof()) {
+        getline(myfile, line);
+        if (line.empty()) {
+          break;
+        }
+        istringstream iss(line);
+        vector<string> tokens;
+        copy(istream_iterator<string>(iss), istream_iterator<string>(),
+          back_inserter(tokens));
+        if (tokens.empty()) {
+          break;
+        } else if (tokens.size() == 1u) {
+          const auto token = tokens[0];
+          if (token == "HOLE") {
+            state = Hole;
+            out_holes.emplace_back();
+            hole = &out_holes.back();
+          } else if (token == "STEINER") {
+            state = Steiner;
+          } else {
+            myfile.close();
+            throw runtime_error("Invalid token [" + token + "]");
+          }
+        } else {
+          double x = StringToDouble(tokens[0]);
+          double y = StringToDouble(tokens[1]);
+          switch (state) {
+          case Polyline:
+            out_polyline.push_back(new Point(x, y));
+            break;
+          case Hole:
+            assert(hole != nullptr);
+            hole->push_back(new Point(x, y));
+            break;
+          case Steiner:
+            out_steiner.push_back(new Point(x, y));
+            break;
+          default:
+            assert(0);
+          }
+        }
+      }
+      myfile.close();
```

Comment:
unneeded as the destructor does this

Reply:
True! and we can remove it also at line 232